File: client/uiwails/services/peers.go
//go:build !(linux && 386)

package services

import (
	"context"
	"fmt"
	"time"

	log "github.com/sirupsen/logrus"

	"github.com/netbirdio/netbird/client/proto"
)

// PeersService exposes peer listing operations to the Wails frontend.
type PeersService struct {
	grpcClient GRPCClientIface
}

// NewPeersService creates a new PeersService.
func NewPeersService(g GRPCClientIface) *PeersService {
	return &PeersService{grpcClient: g}
}

// GetPeers returns the list of all peers with their status information.
func (s *PeersService) GetPeers() ([]PeerInfo, error) {
	conn, err := s.grpcClient.GetClient(3 * time.Second)
	if err != nil {
		log.Debugf("GetPeers: failed to get gRPC client: %v", err)
		return nil, fmt.Errorf("get client: %w", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	resp, err := conn.Status(ctx, &proto.StatusRequest{GetFullPeerStatus: true})
	if err != nil {
		log.Warnf("GetPeers: status RPC failed: %v", err)
		return nil, fmt.Errorf("status rpc: %w", err)
	}

	if resp.FullStatus == nil {
		log.Debugf("GetPeers: fullStatus is nil")
		return []PeerInfo{}, nil
	}

	peers := resp.FullStatus.GetPeers()
	log.Debugf("GetPeers: got %d peers from daemon", len(peers))

	result := make([]PeerInfo, 0, len(peers))
	for _, p := range peers {
		info := PeerInfo{
			IP:               p.GetIP(),
			PubKey:           p.GetPubKey(),
			Fqdn:             p.GetFqdn(),
			ConnStatus:       p.GetConnStatus(),
			Relayed:          p.GetRelayed(),
			RelayAddress:     p.GetRelayAddress(),
			BytesRx:          p.GetBytesRx(),
			BytesTx:          p.GetBytesTx(),
			RosenpassEnabled: p.GetRosenpassEnabled(),
			Networks:         p.GetNetworks(),
			LocalIceType:     p.GetLocalIceCandidateType(),
			RemoteIceType:    p.GetRemoteIceCandidateType(),
			LocalEndpoint:    p.GetLocalIceCandidateEndpoint(),
			RemoteEndpoint:   p.GetRemoteIceCandidateEndpoint(),
		}

		if lat := p.GetLatency(); lat != nil {
			info.LatencyMs = float64(lat.Seconds)*1000 + float64(lat.Nanos)/1e6
		}

		if ts := p.GetLastWireguardHandshake(); ts != nil {
			info.LastHandshake = ts.AsTime().Format(time.RFC3339)
		}

		if ts := p.GetConnStatusUpdate(); ts != nil {
			info.ConnStatusUpdate = ts.AsTime().Format(time.RFC3339)
		}

		result = append(result, info)
	}

	return result, nil
}

// PeerInfo holds simplified peer information for the frontend.
type PeerInfo struct {
	IP               string   `json:"ip"`
	PubKey           string   `json:"pubKey"`
	Fqdn             string   `json:"fqdn"`
	ConnStatus       string   `json:"connStatus"`
	ConnStatusUpdate string   `json:"connStatusUpdate"`
	Relayed          bool     `json:"relayed"`
	RelayAddress     string   `json:"relayAddress"`
	LatencyMs        float64  `json:"latencyMs"`
	BytesRx          int64    `json:"bytesRx"`
	BytesTx          int64    `json:"bytesTx"`
	RosenpassEnabled bool     `json:"rosenpassEnabled"`
	Networks         []string `json:"networks"`
	LastHandshake    string   `json:"lastHandshake"`
	LocalIceType     string   `json:"localIceType"`
	RemoteIceType    string   `json:"remoteIceType"`
	LocalEndpoint    string   `json:"localEndpoint"`
	RemoteEndpoint   string   `json:"remoteEndpoint"`
}

File: client/uitauri/src-tauri/src/commands/peers.rs
use serde::Serialize;
use tauri::State;

use crate::proto;
use crate::state::AppState;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PeerInfo {
    pub ip: String,
    pub pub_key: String,
    pub fqdn: String,
    pub conn_status: String,
    pub conn_status_update: String,
    pub relayed: bool,
    pub relay_address: String,
    pub latency_ms: f64,
    pub bytes_rx: i64,
    pub bytes_tx: i64,
    pub rosenpass_enabled: bool,
    pub networks: Vec<String>,
    pub last_handshake: String,
    pub local_ice_type: String,
    pub remote_ice_type: String,
    pub local_endpoint: String,
    pub remote_endpoint: String,
}

fn format_timestamp(ts: &Option<prost_types::Timestamp>) -> String {
    match ts {
        Some(t) => {
            // Simple RFC3339-like formatting
            let secs = t.seconds;
            let nanos = t.nanos;
            format!("{}:{}", secs, nanos)
        }
        None => String::new(),
    }
}

#[tauri::command]
pub async fn get_peers(state: State<'_, AppState>) -> Result<Vec<PeerInfo>, String> {
    let mut client = state.grpc.get_client().await?;
    let resp = client
        .status(proto::StatusRequest {
            get_full_peer_status: true,
            should_run_probes: false,
            wait_for_ready: None,
        })
        .await
        .map_err(|e| format!("status rpc: {}", e))?
        .into_inner();

    let peers = match resp.full_status {
        Some(ref full) => &full.peers,
        None => return Ok(vec![]),
    };

    let result: Vec<PeerInfo> = peers
        .iter()
        .map(|p| {
            let latency_ms = p
                .latency
                .as_ref()
                .map(|d| d.seconds as f64 * 1000.0 + d.nanos as f64 / 1_000_000.0)
                .unwrap_or(0.0);

            PeerInfo {
                ip: p.ip.clone(),
                pub_key: p.pub_key.clone(),
                fqdn: p.fqdn.clone(),
                conn_status: p.conn_status.clone(),
                conn_status_update: format_timestamp(&p.conn_status_update),
                relayed: p.relayed,
                relay_address: p.relay_address.clone(),
                latency_ms,
                bytes_rx: p.bytes_rx,
                bytes_tx: p.bytes_tx,
                rosenpass_enabled: p.rosenpass_enabled,
                networks: p.networks.clone(),
                last_handshake: format_timestamp(&p.last_wireguard_handshake),
                local_ice_type: p.local_ice_candidate_type.clone(),
                remote_ice_type: p.remote_ice_candidate_type.clone(),
                local_endpoint: p.local_ice_candidate_endpoint.clone(),
                remote_endpoint: p.remote_ice_candidate_endpoint.clone(),
            }
        })
        .collect();

    Ok(result)
}
